File: backend/routes/post.js
const express = require('express');
const router = express.Router();
const db = require('../db');

// 📝 Lấy tất cả post của community user đã tham gia
router.get('/user/:userID', async (req, res) => {
  const { userID } = req.params;
  try {
    const [rows] = await db.query(
      `SELECT 
         p.postID,
         p.userID,
         p.communityID,
         p.postType,
         p.title,
         p.body,
         p.tagList,
         CAST(p.isPinned AS UNSIGNED) AS isPinned,
         CAST(p.viewCount AS UNSIGNED) AS viewCount,
         p.createdAt,
         u.name AS author,
         c.name AS community
       FROM Post p
       JOIN Community c ON p.communityID = c.communityID
       JOIN UserCommunity uc ON c.communityID = uc.communityID
       JOIN User u ON p.userID = u.userID
       WHERE uc.userID = ?
       ORDER BY p.isPinned DESC, p.createdAt DESC`,
      [userID]
    );

    // 🔄 Format createdAt thành ISO string
    const formattedRows = rows.map(row => ({
      ...row,
      createdAt: row.createdAt ? new Date(row.createdAt).toISOString() : null
    }));

    res.json(formattedRows);
  } catch (err) {
    console.error(err);
    res.status(500).json({ error: 'Failed to fetch posts for user' });
  }
});

// 📌 Lấy post theo postID
router.get('/:postID', async (req, res) => {
  try {
    const { postID } = req.params;
    const [rows] = await db.query(
      `SELECT 
         p.postID,
         p.userID,
         p.communityID,
         p.postType,
         p.title,
         p.body,
         p.tagList,
         CAST(p.isPinned AS UNSIGNED) AS isPinned,
         CAST(p.viewCount AS UNSIGNED) AS viewCount,
         p.createdAt,
         u.name AS author,
         c.name AS community
       FROM Post p
       JOIN Community c ON p.communityID = c.communityID
       JOIN User u ON p.userID = u.userID
       WHERE p.postID = ?`,
      [postID]
    );

    if (rows.length === 0) {
      return res.status(404).json({ message: 'Post not found' });
    }

    // Format createdAt
    const post = {
      ...rows[0],
      createdAt: rows[0].createdAt ? new Date(rows[0].createdAt).toISOString() : null
    };

    res.json(post);
  } catch (err) {
    console.error('❌ Error fetching post by ID:', err);
    res.status(500).json({ message: 'Failed to fetch post' });
  }
});

// 📝 Tạo post mới
router.post('/', async (req, res) => {
  console.log('📩 Create Post Body:', req.body);
  const { userID, communityID, postType, title, body, tagList, isPinned = 0, viewCount = 0 } = req.body;

  try {
    const [result] = await db.query(
      `INSERT INTO Post (userID, communityID, postType, title, body, tagList, isPinned, viewCount, createdAt)
       VALUES (?, ?, ?, ?, ?, ?, ?, ?, NOW())`,
      [userID, communityID, postType, title, body, tagList || null, isPinned, viewCount]
    );

    const newPostID = result.insertId;

    // 🔔 Notification broadcast: thêm targetType và targetID
    await db.query(`
      INSERT INTO Notification (senderID, notificationText, targetType, targetID, createdAt)
      SELECT ?, CONCAT(u.name, ' đã đăng bài trong cộng đồng ', c.name, ': ', ?), 'post', ?, NOW()
      FROM User u
      JOIN Community c ON c.communityID = ?
      WHERE u.userID = ?
    `, [userID, title, newPostID, communityID, userID]);

    res.status(201).json({ postID: newPostID });

  } catch (err) {
    console.error('❌ Error creating post:', err);
    res.status(500).json({ error: 'Failed to create post' });
  }
});



module.exports = router;

File: frontend/lib/models/post.dart
import 'package:flutter/material.dart';

class Post {
  final int postID;
  final int userID;
  final int communityID;
  final String postType;
  final String title;
  final String body;
  final String author;
  final String community;
  final DateTime createdAt;
  final bool isPinned;
  final int viewCount;
  final String? tagList;
  Map<String, int> reactions;

  Post({
    required this.postID,
    required this.userID,
    required this.communityID,
    required this.postType,
    required this.title,
    required this.body,
    required this.author,
    required this.community,
    required this.createdAt,
    this.isPinned = false,
    this.viewCount = 0,
    this.tagList,
    Map<String, int>? reactions,
  }) : reactions = reactions ??
            {
              '👍': 0,
              '❤️': 0,
              '😂': 0,
              '😡': 0,
              '😢': 0,
            };

  factory Post.fromJson(Map<String, dynamic> json) {
    debugPrint("🔍 Parsing Post: $json");

    // Mặc định reactions
    Map<String, int> defaultReactions = {
      '👍': 0,
      '❤️': 0,
      '😂': 0,
      '😡': 0,
      '😢': 0,
    };

    try {
      // Parse reactions nếu API trả
      if (json['reactions'] is Map) {
        (json['reactions'] as Map).forEach((key, value) {
          defaultReactions[key] = int.tryParse(value.toString()) ?? 0;
        });
      }

      final post = Post(
        postID: json['postID'] ?? 0,
        userID: json['userID'] ?? 0,
        communityID: json['communityID'] ?? 0,
        postType: json['postType']?.toString() ?? '',
        title: json['title']?.toString() ?? '',
        body: json['body']?.toString() ?? '',
        author: json['author']?.toString() ?? 'Unknown',
        community: json['community']?.toString() ?? '',
        createdAt: DateTime.tryParse(json['createdAt']?.toString() ?? '') ??
            DateTime.now(),
        isPinned: json['isPinned'] == 1 || json['isPinned'] == true,
        viewCount: int.tryParse(json['viewCount']?.toString() ?? '0') ?? 0,
        tagList: json['tagList']?.toString(),
        reactions: defaultReactions,
      );

      debugPrint(
          "✅ Parsed Post: ${post.postID} - ${post.title} (${post.createdAt.toIso8601String()})");
      return post;
    } catch (e) {
      debugPrint("❌ Lỗi parse Post: $e | Data: $json");
      return Post(
        postID: 0,
        userID: 0,
        communityID: 0,
        postType: '',
        title: '',
        body: '',
        author: '',
        community: '',
        createdAt: DateTime.now(),
        reactions: defaultReactions,
      );
    }
  }

  Map<String, dynamic> toJson() {
    return {
      'postID': postID,
      'userID': userID,
      'communityID': communityID,
      'postType': postType,
      'title': title,
      'body': body,
      'tagList': tagList,
      'isPinned': isPinned ? 1 : 0,
      'viewCount': viewCount,
      'author': author,
      'community': community,
      'createdAt': createdAt.toIso8601String(),
      'reactions': reactions,
    };
  }
}
